include/networkit/graph/Graph.hpp

```diff
@@ -139,6 +145,389 @@ class Graph final {
     //!< same schema (and same order!) as outEdges
     std::vector<std::vector<edgeid>> outEdgeIds;
 
+private:
+    // user attributes of this graph (node attributes so far)
+    // Base class for all node attributes.
+    class NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorageBase(const Graph *graph, std::string name, std::type_index type)
+            : name{std::move(name)}, type{type}, theGraph{graph} {}
+
+        virtual ~NodeAttributeStorageBase() = default;
+
+        virtual void invalidateAttributes() = 0;
+
+        std::string getName() { return name; }
+
+        std::type_index getType() { return type; }
+
+        bool isValid(index i) { return i < valid.size() && valid[i]; }
+
+        // Called by Graph when node is deleted.
+        void invalidate(index i) {
+            if (isValid(i)) {
+                valid[i] = false;
+                --validElements;
+            }
+        }
+
+    protected:
+        void markValid(index i) {
+            if (!theGraph->hasNode(i)) {
+                throw std::runtime_error("This node does not exist");
+            }
+            if (i >= valid.size()) {
+                valid.resize(i + 1);
+            }
+            valid[i] = true;
+            ++validElements;
+        }
+
+        void checkIndex(index i) {
+            if (!theGraph->hasNode(i)) {
+                throw std::runtime_error("This node does not exist");
+            }
+            if (!isValid(i)) {
+                throw std::runtime_error("Invalid attribute value");
+            }
```

Comment:
Same here. It seems to me that this check is useful for debugging, so we should use assertions instead of runtime checks.